src/Features/CSharp/Portable/UseExpressionBodyForLambda/UseExpressionBodyForLambdaCodeStyleProvider.cs

```diff
@@ -0,0 +1,436 @@
+            }
+
+            return updatedDecl;
+        }
+
+        private static bool TryConvertToExpressionBody(
+            LambdaExpressionSyntax declaration, ParseOptions options, ExpressionBodyPreference conversionPreference,
+            out ExpressionSyntax expression, out SyntaxToken semicolon)
+        {
+            var body = declaration.Body as BlockSyntax;
+
+            return body.TryConvertToExpressionBody(
+                declaration.Kind(), options, conversionPreference,
+                out expression, out semicolon);
+        }
+
+        private static LambdaExpressionSyntax WithBlockBody(
+            SemanticModel semanticModel, LambdaExpressionSyntax declaration)
+        {
+            var expressionBody = GetBodyAsExpression(declaration);
+            var lambdaType = (INamedTypeSymbol)semanticModel.GetTypeInfo(declaration).ConvertedType;
+            var createReturnStatementForExpression = !lambdaType.DelegateInvokeMethod.ReturnsVoid;
+
+            if (!expressionBody.TryConvertToStatement(
+                    semicolonTokenOpt: default,
+                    createReturnStatementForExpression,
+                    out var statement))
+            {
+
+                return declaration;
+            }
+
+            // If the user is converting to a block, it's likely they intend to add multiple
+            // statements to it.  So make a multi-line block so that things are formatted properly
+            // for them to do so.
+            return declaration.WithBody(SyntaxFactory.Block(
+                SyntaxFactory.Token(SyntaxKind.OpenBraceToken).WithAppendedTrailingTrivia(SyntaxFactory.ElasticCarriageReturnLineFeed),
+                SyntaxFactory.SingletonList(statement),
+                SyntaxFactory.Token(SyntaxKind.CloseBraceToken)));
+        }
+
+        #endregion
+
+        private class MyCodeAction : CodeAction.DocumentChangeAction
+        {
+            public MyCodeAction(string title, Func<CancellationToken, Task<Document>> createChangedDocument)
+                : base(title, createChangedDocument)
+            {
+            }
+        }
+    }
+
+    // Stub classes needed only for exporting purposes.
+
+    [ExportCodeFixProvider(LanguageNames.CSharp), Shared]
+    internal partial class UseExpressionBodyForLambdaCodeFixProvider : UseExpressionBodyForLambdaCodeStyleProvider.CodeFixProvider
+    {
+    }
+
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp,
+        Name = nameof(UseExpressionBodyForLambdaCodeRefactoringProvider)), Shared]
+    internal class UseExpressionBodyForLambdaCodeRefactoringProvider : UseExpressionBodyForLambdaCodeStyleProvider.CodeRefactoringProvider
+    {
+    }
+
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal class UseExpressionBodyForLambdaDiagnosticAnalyzer : UseExpressionBodyForLambdaCodeStyleProvider.DiagnosticAnalyzer
```

Comment:
These are the three classes a feature would still have to export. Note they are very intentionally empty. They do not need to do anything, and literally exist solely so we can add the export attributes somewhere so these can be found by the relevant subsystems.

in an ideal world we could just put these attributes on the CodeStyleProvider. however, since each of these attributes expects a particular inheritance hierarchy, and because they all use classes, there is no way to make CodeStyleProvider extend all three of these classes.

Alternatively, we could just have something like ExportCodeStyleProvider, and have the relevant subsystems know how to import that. Then, the CodeStyleProvider would expose the 3 separate providers from itself. That would be trivial to do for CodeRefactoringProvider/CodeFixProvider. however, it's unclear to me how we would fit that into the DiagnosticAnalyzer infrastructure.